common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java

```diff
@@ -71,32 +72,49 @@
      * @param executor           An Executor to use for callback invocations.
      */
     public static void onStop(Service service, Runnable terminatedCallback, Consumer<Throwable> failureCallback, Executor executor) {
+        ShutdownListener listener = new ShutdownListener(terminatedCallback, failureCallback);
+        service.addListener(listener, executor);
+
+        // addListener() will not invoke the callbacks if the service is already in a terminal state. As such, we need to
```

Comment:
I don't follow this change compared to the description of the pull request. The description says:

```
Fixed a bug in ServiceHelpers where it could be possible that callbacks passed to onStop() would never execute in certain cases.
```

but the logic here compared to what is currently in master seems to guarantee no multiple calls to the callbacks, not that they will be invoked at least once. What am I missing?